add noninitial plastic strain property

naliboff opened the pull request:
This pull request enables the visco plastic strain invariant particle property to track noninitial plastic strain, and also (hopefully) improves the code structure and documentation. I tested this locally with a new test case, but we will see what happens with the existing tests.

* [X] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [X] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

naliboff commented:
/rebuild

anne-glerum commented on source/particle/property/viscoplastic_strain_invariants.cc:
```diff
@@ -123,39 +129,46 @@ namespace aspect
-          data[data_position] = new_strain;
+          data[data_position] += strain_update;
 
         if (this->introspection().compositional_name_exists("viscous_strain") && plastic_yielding == false)
```
Thank you for adding the list of assumptions above, it's much clearer! Could you please also add for each of the below conditions a comment that explains what situation we're in? I'll add some examples below. I think that will make it easier to change the code if we ever change the assumptions.

anne-glerum commented on source/particle/property/viscoplastic_strain_invariants.cc:
```diff
@@ -123,39 +129,46 @@ namespace aspect
         if (this->introspection().compositional_name_exists("viscous_strain") && plastic_yielding == false)
-          data[data_position+(n_components-1)] = new_strain;
+          {
+            if (n_components == 1)
```
`// Not yielding and only one field, which tracks the viscous strain`

anne-glerum commented on source/particle/property/viscoplastic_strain_invariants.cc:
```diff
@@ -123,39 +129,46 @@ namespace aspect
+            if (n_components == 1)
+              data[data_position] += strain_update;
+
+            if (n_components == 2)
```
`// Not yielding, two fields, second is viscous strain.`

anne-glerum commented on source/particle/property/viscoplastic_strain_invariants.cc:
```diff
@@ -123,39 +129,46 @@ namespace aspect
+            if (n_components == 2)
+              data[data_position+(n_components-1)] += strain_update;
+
+            if (n_components == 3)
```
`// Not yielding, three fields (plastic strain, viscous strain, nonintial plastic strain).`

anne-glerum reviewed (changes requested):
Thank you John, that's a lot clearer! Can you add a test that tracks the noninitial strain on particles?

naliboff commented:
@anne-glerum - Thanks for the review and I've hopefully addressed all of the comments. Good idea to add a description above each scenario, which will make this much easier to update later on. Please let me know if any additional changes are needed.

anne-glerum commented on tests/visco_plastic_yield_noninitial-plastic-strain_particles.prm:
```diff
@@ -0,0 +1,51 @@
+# A test to check whether noninitial plastic strain is correctly tracked
+# on particles, which are also used to track plastic and viscous strain.
+# The boundary conditions (see tests/visco_plastic_yield_plastic_viscous_strain_weakening.prm)
+# produce a incremental strain of 1e-5 per time step and material that is always yielding.
```
```suggestion
# produce an incremental strain of 1e-5 per time step and material that is always yielding.
```

anne-glerum reviewed (changes requested):
Thanks for adding the test! Could you please add a line about the noninitial_plastic_strain in the description of the particle property at the bottom of the cc file? Sorry I didn't see that before.

naliboff commented on tests/visco_plastic_yield_noninitial-plastic-strain_particles.prm:
```diff
@@ -0,0 +1,51 @@
+set Maximum time step  = 1000
+set Output directory   = visco_plastic_yield_noninitial-plastic-strain_particles
+
+# Particles mapped to the Compositional fields used to track plastic, noninitial plastic, and viscous strain.
```
Yep, it took me a bit of looking to figure out this out as well!

naliboff commented:
@anne-glerum - Thanks for the additional review and suggestions! I've incorporated all of the suggestions and also simplified the logic in the viscous strain update section, per your recommendation. Assuming the tests still pass, let me know if you would like any additional changes.

anne-glerum reviewed (approved):
Great thank you!

anne-glerum commented:
Do you think this warrants a change log entry?

naliboff commented:
@anne-glerum - Good point, I added a change log entry.

anne-glerum commented:
Thanks John!
